### PR TITLE
DUPLO-14936, DUPLO-14934, DUPLO-14935

### PR DESCRIPTION
**Pull request by michalpleszkun**
## Overview

- Added example of usage for EFS lifecycles policies
- Removed transision_to_archive lifecycle policy

## Summary of changes

This PR does the following:

- ...
- ...

## Testing performed

- [ ] Using unit tests
- [x] Manually, on my local system
- [ ] Manually, on a remote test system

## Describe any breaking changes

- ...

___

## **Type**
enhancement, documentation

___

## **Description**
- Enhanced EFS lifecycle policies with detailed descriptions for transition rules.
- Made `lifecycle_policy` a required field in `resource_duplo_aws_efs_file_system_lifecycle_policy.go`.
- Corrected the `FileSystemId` assignment in the EFS update function to use `efsId`.
- Added Terraform and shell script examples for creating and importing EFS lifecycle policies.
- Updated the `LifecyclePolicy` struct in `duplosdk` to use `DuploStringValue` for transition attributes.
- Updated CHANGELOG.md with details of the enhancements and fixes.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>resource_duplo_aws_efs_file_system.go</strong><dd><code>Enhance EFS Lifecycle Policies and Fix `FileSystemId` Assignment</code></dd></summary>
<hr>
      
duplocloud/resource_duplo_aws_efs_file_system.go

<li>Added detailed descriptions for <code>transition_to_archive</code>, <br><code>transition_to_ia</code>, and <code>transition_to_primary_storage_class</code> in EFS <br>lifecycle policies.<br> <li> Corrected the <code>FileSystemId</code> assignment in the EFS update function.<br>

</details>
    

  </td>
  <td><a href="https:/duplocloud/terraform-provider-duplocloud/pull/412/files#diff-7d78cc8cb996471fa7b6797110f58dd9a8363df9a5a7eaa7637ef75edc75878b">+4/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource_duplo_aws_efs_file_system_lifecycle_policy.go</strong><dd><code>Update Lifecycle Policy Schema and ID Handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplocloud/resource_duplo_aws_efs_file_system_lifecycle_policy.go

<li>Added detailed descriptions for lifecycle policy fields.<br> <li> Changed <code>lifecycle_policy</code> from optional to required.<br> <li> Implemented ID formatting and parsing for lifecycle policies.<br>

</details>
    

  </td>
  <td><a href="https:/duplocloud/terraform-provider-duplocloud/pull/412/files#diff-48c85f23c9df45e3604321cdfea72646c9127e7fff6ca104ea4e51860f13ba50">+29/-9</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>aws_efs_lifecycle_policy.go</strong><dd><code>Refactor LifecyclePolicy Struct to Use DuploStringValue</code>&nbsp; &nbsp; </dd></summary>
<hr>
      
duplosdk/aws_efs_lifecycle_policy.go

<li>Updated <code>LifecyclePolicy</code> struct to use <code>DuploStringValue</code> for transition <br>attributes.<br>

</details>
    

  </td>
  <td><a href="https:/duplocloud/terraform-provider-duplocloud/pull/412/files#diff-845de39655ee678d5bcf91fd14ec9a935bf5cd0c43c6fb0747d1467a472a5b8b">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>import.sh</strong><dd><code>Add Import Example for AWS EFS Lifecycle Policy</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
examples/resources/duplocloud_aws_efs_lifecycle_policy/import.sh

<li>Added an example script for importing an existing AWS EFS lifecycle <br>policy.<br>

</details>
    

  </td>
  <td><a href="https:/duplocloud/terraform-provider-duplocloud/pull/412/files#diff-60a7e0ff03091f7cbbc671a8bb445572c01a86b97d3ae0a5ecf2733edb47dca7">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource.tf</strong><dd><code>Add Terraform Example for EFS Lifecycle Policy</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf

<li>Provided a Terraform configuration example for creating an EFS <br>lifecycle policy.<br>

</details>
    

  </td>
  <td><a href="https:/duplocloud/terraform-provider-duplocloud/pull/412/files#diff-f1003862bfd9cbf94c12d62cc17c87f4315fe948c3f28467e7d01eaade7dcc47">+27/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Update Changelog with EFS Lifecycle Policy Enhancements</code>&nbsp; &nbsp; </dd></summary>
<hr>
      
CHANGELOG.md

<li>Documented enhancements and fixes related to EFS lifecycle policies.<br> <li> Added entries for other features and fixes unrelated to this PR for <br>context.<br>

</details>
    

  </td>
  <td><a href="https:/duplocloud/terraform-provider-duplocloud/pull/412/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+6/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-14936 TF:efs_lifecycle_policy: getting error "One or more LifecyclePolicy objects specified are malformed" for `transition_to_archive` lifecycle policy ](https://app.clickup.com/t/8655600/DUPLO-14936)

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-14934 TF Document changes required for  `duplocloud_aws_efs_lifecycle_policy` resource](https://app.clickup.com/t/8655600/DUPLO-14934)

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,15 @@
+## 2024-02-16
+
+### Changed
+- Enhanced EFS lifecycle policies by adding detailed descriptions for `transition_to_ia` and `transition_to_primary_storage_class` policies.
```
Comment (joek-duplo) [inline CHANGELOG.md:4]: TF provider changelog entries must have the TF resource name.  Please see the other entries for examples.
Also [see here](https://github.com/duplocloud-internal/wiki/wiki/ProductDevBranchAndReviewModel#changelogmd-maintenance-guidelines-and-requirements).

File: duplocloud/resource_duplo_aws_efs_file_system.go [hunk 2]
```diff
@@ -45,20 +45,17 @@ func awsEFSFileSystem() map[string]*schema.Schema {
 		"lifecycle_policy": {
 			Type:     schema.TypeList,
 			Optional: true,
-			MaxItems: 3,
+			MaxItems: 2,
 			Elem: &schema.Resource{
 				Schema: map[string]*schema.Schema{
-					"transition_to_archive": {
```
Comment (joek-duplo) [inline duplocloud/resource_duplo_aws_efs_file_system.go:51]: I don't think we can remove `transition_to_archive`, I believe this must continue to be supported.

File: examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  backup           = true
+  encrypted        = true
+  provisioned_throughput_in_mibps= 20
+}
+
+resource "duplocloud_aws_efs_lifecycle_policy" "efs_policy" {
+  tenant_id        = duplocloud_tenant.myapp.tenant_id
+  file_system_id             = duplocloud_aws_efs_file_system.efs.file_system_id
+  lifecycle_policy{
+  transition_to_ia = "AFTER_7_DAYS"
```
Comment (joek-duplo) [inline examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf:20]: code format:  fix indentation

File: examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  }
+  lifecycle_policy{
+  transition_to_primary_storage_class = "AFTER_1_ACCESS"
```
Comment (joek-duplo) [inline examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf:23]: code format:  fix indentation

File: examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf [hunk 5]
```diff
@@ -0,0 +1,26 @@
+resource "duplocloud_tenant" "myapp" {
```
Comment (joek-duplo) [inline examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf:1]: we should also add an `import.sh` script file in addition to this `resource.tf` file

**Review by joek-duplo — CHANGES_REQUESTED** [review]

File: duplocloud/resource_duplo_aws_efs_file_system.go (hunk 2 above)
Comment (michalpleszkun) [reply duplocloud/resource_duplo_aws_efs_file_system.go:51]: We can and we must. We are not able to set it and this is the reason of DUPLO-14936.

Comment (joek-duplo) [reply duplocloud/resource_duplo_aws_efs_file_system.go:51]: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/efs_file_system#transition_to_archive

Comment (joek-duplo) [reply duplocloud/resource_duplo_aws_efs_file_system.go:51]: why can AWS support this field, but Duplo cannot?

**Review by joek-duplo — CHANGES_REQUESTED** [review]

File: examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf [hunk 6]
```diff
@@ -18,7 +18,10 @@ resource "duplocloud_aws_efs_lifecycle_policy" "efs_policy" {
   lifecycle_policy {
     transition_to_ia = "AFTER_7_DAYS"
   }
+  lifecycle_policy{
+  transition_to_archive = "AFTER_14_DAYS"
```
Comment (joek-duplo) [inline examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf:22]: fix indentation

File: examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf [hunk 7]
```diff
@@ -18,7 +18,10 @@ resource "duplocloud_aws_efs_lifecycle_policy" "efs_policy" {
   lifecycle_policy {
     transition_to_ia = "AFTER_7_DAYS"
   }
+  lifecycle_policy{
```
Comment (joek-duplo) [inline examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf:21]: fix missing space before bracket

**Review by joek-duplo — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (michalpleszkun) [reply CHANGELOG.md]: I think it is done.

File: examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf (hunk 4 above)
Comment (michalpleszkun) [reply examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf]: done

File: examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf (hunk 6 above)
Comment (michalpleszkun) [reply examples/resources/duplocloud_aws_efs_lifecycle_policy/resource.tf]: done

**Review by joek-duplo — CHANGES_REQUESTED** [review]
This PR is still removing the `transition_to_archive` element - so I can't approve it yet.

**Review by joek-duplo — APPROVED** [review]

**Review by joek-duplo — APPROVED** [review]

**Pull request merged**